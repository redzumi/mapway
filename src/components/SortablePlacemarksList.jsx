import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styled from 'styled-components'
import { arrayMove, SortableContainer, SortableElement, SortableHandle } from 'react-sortable-hoc'

const DragHandle = SortableHandle(() => (
  <span>
    <ReorderIcon className="material-icons">reorder</ReorderIcon>
  </span>
))

const SortableItem = SortableElement(({ value, onDelete }) => (
  <StyledLi>
    <DragHandle />
    {value}
    <DeleteIcon
      onClick={onDelete}
      onKeyPress={onDelete}
      role="button"
      tabIndex="0"
      className="material-icons delete-icon">
      delete
    </DeleteIcon>
  </StyledLi>
))

const SortableList = SortableContainer(({ items, onItemDelete }) => (
  <StyledUl>
    {items.map((value, index) => (
      <SortableItem
        key={value.id}
        index={index}
        value={value.name}
        onDelete={() => onItemDelete(value)}
      />
    ))}
  </StyledUl>
))

class SortablePlacemarksList extends Component {
  constructor(props) {
    super(props)
    this.state = { placemarks: this.props.placemarks }
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.placemarks !== nextProps.placemarks)
      this.setState({ placemarks: nextProps.placemarks })
  }

  handleSortEnd = ({ oldIndex, newIndex }) => {
    const placemarks = arrayMove(this.state.placemarks, oldIndex, newIndex)
    this.props.onPlacemarksOrderUpdate(placemarks)
    this.setState({ placemarks })
  }

  handleItemDelete = item => {
    this.props.onPlacemarkDelete(this.state.placemarks.indexOf(item))
  }

  render() {
    return (
      <SortableList
        lockAxis="y"
        lockToContainerEdges
        useDragHandle
        items={this.state.placemarks}
        onItemDelete={this.handleItemDelete}
        onSortEnd={this.handleSortEnd}
      />
    )
  }
}

SortablePlacemarksList.propTypes = {
  placemarks: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      coordinates: PropTypes.array,
    })
  ).isRequired,
  onPlacemarksOrderUpdate: PropTypes.func.isRequired,
  onPlacemarkDelete: PropTypes.func.isRequired,
}

const DeleteIcon = styled.i`
  cursor: pointer;
`

const ReorderIcon = styled.i`
  cursor: row-resize;
`

const StyledLi = styled.li`
  height: 60px;
  width: auto;
  padding: 0px 10px;
  background-color: #fff;
  display: flex;
  justify-content: space-between;
  font-family: 'Montserrat', sans-serif;
  align-items: center;
  border-bottom: 1px solid #efefef;
  box-sizing: border-box;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.2);
`

const StyledUl = styled.ul`
  background-color: #f3f3f3;
  border: 1px solid #efefef;
  width: 300px;
  padding: 0px;
  margin: 0px;
  outline: none;
  user-select: none;
  & li {
    box-shadow: none;
  }
  &:active {
    cursor: row-resize;
  }
`

export { SortablePlacemarksList as default, StyledLi, StyledUl }
