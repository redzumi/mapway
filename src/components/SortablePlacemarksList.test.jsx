import React from 'react'
import { mount, shallow } from 'enzyme'
import SortablePlacemarksList, { StyledLi, StyledUl } from './SortablePlacemarksList'

const onPlacemarksOrderUpdate = jest.fn()
const onPlacemarkDelete = jest.fn()
const placemarks = [
  {
    id: 'id1',
    name: 'Name 1',
    coordinates: [55, 55],
  },
  {
    id: 'id2',
    name: 'Name 2',
    coordinates: [55, 55],
  },
]

describe('SortablePlacemarksList styled components', () => {
  it('renders ul correctly', () => {
    expect(shallow(<StyledUl />)).toMatchSnapshot()
  })

  it('renders li correctly', () => {
    expect(shallow(<StyledLi />)).toMatchSnapshot()
  })
})

describe('SortablePlacemarksList', () => {
  let renderedComponent
  beforeEach(() => {
    renderedComponent = mount(
      <SortablePlacemarksList
        placemarks={placemarks}
        onPlacemarksOrderUpdate={onPlacemarksOrderUpdate}
        onPlacemarkDelete={onPlacemarkDelete}
      />
    )
  })

  afterEach(() => {
    onPlacemarksOrderUpdate.mockReset()
    onPlacemarkDelete.mockReset()
  })

  it('renders correctly', () => {
    expect(renderedComponent).toMatchSnapshot()
  })

  it('reorders placemarks', () => {
    renderedComponent.instance().handleSortEnd({ oldIndex: 0, newIndex: 1 })

    expect(onPlacemarksOrderUpdate).toBeCalledWith([
      {
        id: 'id2',
        name: 'Name 2',
        coordinates: [55, 55],
      },
      {
        id: 'id1',
        name: 'Name 1',
        coordinates: [55, 55],
      },
    ])
  })

  it('deletes placemark from list', () => {
    renderedComponent
      .find('.delete-icon')
      .first()
      .simulate('click')

    expect(onPlacemarkDelete).toBeCalledWith(0)
  })
})
