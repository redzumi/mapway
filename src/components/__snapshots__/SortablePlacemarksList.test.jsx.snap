// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortablePlacemarksList renders correctly 1`] = `
.c3 {
  cursor: pointer;
}

.c2 {
  cursor: row-resize;
}

.c1 {
  height: 60px;
  width: auto;
  padding: 0px 10px;
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-family: 'Montserrat',sans-serif;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-bottom: 1px solid #efefef;
  box-sizing: border-box;
  box-shadow: 0 2px 10px rgba(0,0,0,0.2);
}

.c0 {
  background-color: #f3f3f3;
  border: 1px solid #efefef;
  width: 300px;
  padding: 0px;
  margin: 0px;
  outline: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 li {
  box-shadow: none;
}

.c0:active {
  cursor: row-resize;
}

<SortablePlacemarksList
  onPlacemarkDelete={[MockFunction]}
  onPlacemarksOrderUpdate={[MockFunction]}
  placemarks={
    Array [
      Object {
        "coordinates": Array [
          55,
          55,
        ],
        "id": "id1",
        "name": "Name 1",
      },
      Object {
        "coordinates": Array [
          55,
          55,
        ],
        "id": "id2",
        "name": "Name 2",
      },
    ]
  }
>
  <sortableList
    axis="y"
    distance={0}
    getHelperDimensions={[Function]}
    hideSortableGhost={true}
    items={
      Array [
        Object {
          "coordinates": Array [
            55,
            55,
          ],
          "id": "id1",
          "name": "Name 1",
        },
        Object {
          "coordinates": Array [
            55,
            55,
          ],
          "id": "id2",
          "name": "Name 2",
        },
      ]
    }
    lockAxis="y"
    lockOffset="50%"
    lockToContainerEdges={true}
    onItemDelete={[Function]}
    onSortEnd={[Function]}
    pressDelay={0}
    pressThreshold={5}
    shouldCancelStart={[Function]}
    transitionDuration={300}
    useDragHandle={true}
    useWindowAsScrollContainer={false}
  >
    <Component
      items={
        Array [
          Object {
            "coordinates": Array [
              55,
              55,
            ],
            "id": "id1",
            "name": "Name 1",
          },
          Object {
            "coordinates": Array [
              55,
              55,
            ],
            "id": "id2",
            "name": "Name 2",
          },
        ]
      }
      onItemDelete={[Function]}
    >
      <styled.ul>
        <ul
          className="c0"
        >
          <sortableElement
            collection={0}
            index={0}
            key="id1"
            onDelete={[Function]}
            value="Name 1"
          >
            <Component
              onDelete={[Function]}
              value="Name 1"
            >
              <styled.li>
                <li
                  className="c1"
                >
                  <sortableHandle>
                    <Component>
                      <span>
                        <styled.i
                          className="material-icons"
                        >
                          <i
                            className="material-icons c2"
                          >
                            reorder
                          </i>
                        </styled.i>
                      </span>
                    </Component>
                  </sortableHandle>
                  Name 1
                  <styled.i
                    className="material-icons delete-icon"
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    role="button"
                    tabIndex="0"
                  >
                    <i
                      className="material-icons delete-icon c3"
                      onClick={[Function]}
                      onKeyPress={[Function]}
                      role="button"
                      tabIndex="0"
                    >
                      delete
                    </i>
                  </styled.i>
                </li>
              </styled.li>
            </Component>
          </sortableElement>
          <sortableElement
            collection={0}
            index={1}
            key="id2"
            onDelete={[Function]}
            value="Name 2"
          >
            <Component
              onDelete={[Function]}
              value="Name 2"
            >
              <styled.li>
                <li
                  className="c1"
                >
                  <sortableHandle>
                    <Component>
                      <span>
                        <styled.i
                          className="material-icons"
                        >
                          <i
                            className="material-icons c2"
                          >
                            reorder
                          </i>
                        </styled.i>
                      </span>
                    </Component>
                  </sortableHandle>
                  Name 2
                  <styled.i
                    className="material-icons delete-icon"
                    onClick={[Function]}
                    onKeyPress={[Function]}
                    role="button"
                    tabIndex="0"
                  >
                    <i
                      className="material-icons delete-icon c3"
                      onClick={[Function]}
                      onKeyPress={[Function]}
                      role="button"
                      tabIndex="0"
                    >
                      delete
                    </i>
                  </styled.i>
                </li>
              </styled.li>
            </Component>
          </sortableElement>
        </ul>
      </styled.ul>
    </Component>
  </sortableList>
</SortablePlacemarksList>
`;

exports[`SortablePlacemarksList styled components renders li correctly 1`] = `
.c0 {
  height: 60px;
  width: auto;
  padding: 0px 10px;
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-family: 'Montserrat',sans-serif;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-bottom: 1px solid #efefef;
  box-sizing: border-box;
  box-shadow: 0 2px 10px rgba(0,0,0,0.2);
}

<li
  className="c0"
/>
`;

exports[`SortablePlacemarksList styled components renders ul correctly 1`] = `
.c0 {
  background-color: #f3f3f3;
  border: 1px solid #efefef;
  width: 300px;
  padding: 0px;
  margin: 0px;
  outline: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 li {
  box-shadow: none;
}

.c0:active {
  cursor: row-resize;
}

<ul
  className="c0"
/>
`;
